### PR TITLE
TYP: first_valid_index & last_valid_index

### PR DESCRIPTION
**Pull request by topper-123**
Typing of `first_valid_index` & `last_valid_index` + make the `how` param in internal functions a named parameter for better readability.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -11143,7 +11143,7 @@ def __ixor__(self, other):
     # Misc methods
 
     @final
-    def _find_valid_index(self, how: str):
+    def _find_valid_index(self, *, how: str) -> Optional[Hashable]:
```
Comment (jbrockmendel) [inline pandas/core/generic.py:11146]: just Hashable? xref #40534

Comment (topper-123) [reply pandas/core/generic.py:11146]: I commented on that here in in #40534: https://github.com/pandas-dev/pandas/pull/40534#issuecomment-804010959.

I don't have a definite opinion, but this is the same as in many other cases in the code base and emphasizes that None is special, i.e. in `df[df.find_first_valid(): df.find_last_valid()]` a return value in `_find_valid_index` of None means slice to the start/end and not to that value.

If this should be just `Hashable`, we should probably do that in the whole code base.

Comment (jbrockmendel) [reply pandas/core/generic.py:11146]: make sense

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**